src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,422 @@
+
+variables {F Γ k}
+
+instance : has_coe_to_fun (modular_form Γ k) (λ _, ℍ → ℂ) := fun_like.has_coe_to_fun
+instance : has_coe_to_fun (cusp_form Γ k) (λ _, ℍ → ℂ) := fun_like.has_coe_to_fun
+
+@[simp] lemma modular_form_to_fun_eq_coe {f : modular_form Γ k} : f.to_fun = (f : ℍ → ℂ) := rfl
+@[simp] lemma cusp_form_to_fun_eq_coe {f : cusp_form Γ k} : f.to_fun = (f : ℍ → ℂ) := rfl
+
+@[ext] theorem modular_form_ext {f g : modular_form Γ k} (h : ∀ x, f x = g x) : f = g :=
+fun_like.ext f g h
+
+@[ext] theorem cusp_form_ext {f g : cusp_form Γ k} (h : ∀ x, f x = g x) : f = g :=
+fun_like.ext f g h
+
+/-- Copy of a `modular_form` with a new `to_fun` equal to the old one. Useful to fix
+definitional equalities. -/
+protected def modular_form_copy (f : modular_form Γ k) (f' : ℍ → ℂ) (h : f' = ⇑f) :
+  modular_form Γ k :=
+{ to_fun := f',
+  slash_action_eq' := h.symm ▸ f.slash_action_eq',
+  hol' := h.symm ▸ f.hol',
+  bdd_at_infty' := λ A, h.symm ▸ f.bdd_at_infty' A }
+
+/-- Copy of a `cusp_form` with a new `to_fun` equal to the old one. Useful to fix
+definitional equalities. -/
+protected def cusp_form_copy (f : cusp_form Γ k) (f' : ℍ → ℂ) (h : f' = ⇑f) :
+  cusp_form Γ k :=
+{ to_fun := f',
+  slash_action_eq' := h.symm ▸ f.slash_action_eq',
+  hol' := h.symm ▸ f.hol',
+  zero_at_infty' := λ A, h.symm ▸ f.zero_at_infty' A }
+
+end modular_form
+
+namespace modular_form
+
+open slash_invariant_forms
+
+variables {F : Type*} {Γ : subgroup SL(2, ℤ)} {k : ℤ}
+
+instance has_add : has_add (modular_form Γ k) :=
+{ add := λ f g,
+  { to_fun := f + g,
+    slash_action_eq' := (f.to_slash_invariant_form + g.to_slash_invariant_form).2,
+    hol' := f.hol'.add _ g.hol',
+    bdd_at_infty' := λ A, begin
+        rw slash_action.add_action,
+        exact (bounded_at_im_infty_subalgebra ℂ).add_mem' (f.bdd_at_infty' A) (g.bdd_at_infty' A),
+        end}}
+
+@[simp] lemma coe_add {f g : (modular_form Γ k)} : ⇑(f + g) = f + g := rfl
+@[simp] lemma add_apply {f g : (modular_form Γ k)} {z : ℍ} : (f + g) z = f z + g z := rfl
+
+instance has_zero : has_zero (modular_form Γ k) :=
+{ zero := ⟨0, slash_action.mul_zero _, (λ _, mdifferentiable_at_const 𝓘(ℂ, ℂ) 𝓘(ℂ, ℂ)),
+  by {intro a,
+    convert (bounded_at_im_infty_subalgebra ℂ).zero_mem',
+    apply slash_action.mul_zero _ }⟩}
+
+instance has_nsmul : has_smul ℕ (modular_form Γ k) :=
+⟨ λ c f,
+  { to_fun := (c : ℂ) • f,
+    slash_action_eq' := by {intro γ, convert slash_action.smul_action k γ f (c : ℂ),
+    exact ((f.slash_action_eq') γ).symm},
+    hol' := f.hol'.const_smul _ _,
+    bdd_at_infty' := λ A, begin
+      rw slash_action.smul_action,
+      apply (bounded_at_im_infty_subalgebra ℂ).smul_mem (f.bdd_at_infty' A)
+    end }⟩
+
+@[simp] lemma coe_nsmul {f : (modular_form Γ k)} {n : ℕ} : ⇑(n • f) = n • f :=
+begin
+  simp only [nsmul_eq_mul],
+  refl,
+end
+
+@[simp] lemma nsmul_apply {f : (modular_form Γ k)} {n : ℕ} {z : ℍ} :
+   (n • f) z = (n : ℂ) • (f z) := rfl
```

Comment:
This is inconsistent with `coe_nsmul` since you're not coercing the `n` to complex on that lemma.  It probably shouldn't matter; I think the simp paths will end up in the same place.  But one shouldn't need both; for conciseness, can you try simply deleting this lemma?